In [13]:
import pandas as pd
import numpy as np

In [14]:
df = pd.read_csv('PlayerStatistics.csv')

C:\Users\szhak\AppData\Local\Temp\ipykernel_10028\2152893101.py:1: DtypeWarning: Columns (0: gameLabel, 1: gameSubLabel, 2: seriesGameNumber, 3: numMinutes, 4: comment, 5: startingPosition) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('PlayerStatistics.csv')


In [15]:
print('Строк и колонок:', df.shape)

Строк и колонок: (1669922, 40)


In [16]:
print('Названия колонок:')
print(df.columns.tolist())

Названия колонок:
['firstName', 'lastName', 'personId', 'gameId', 'gameDateTimeEst', 'playerteamCity', 'playerteamName', 'opponentteamCity', 'opponentteamName', 'gameType', 'gameLabel', 'gameSubLabel', 'seriesGameNumber', 'win', 'home', 'numMinutes', 'points', 'assists', 'blocks', 'steals', 'fieldGoalsAttempted', 'fieldGoalsMade', 'fieldGoalsPercentage', 'threePointersAttempted', 'threePointersMade', 'threePointersPercentage', 'freeThrowsAttempted', 'freeThrowsMade', 'freeThrowsPercentage', 'reboundsDefensive', 'reboundsOffensive', 'reboundsTotal', 'foulsPersonal', 'turnovers', 'plusMinusPoints', 'playerteamId', 'opponentteamId', 'comment', 'startingPosition', 'gameDate']


In [17]:
print('Типы данных:')
print(df.dtypes)

Типы данных:
firstName                      str
lastName                       str
personId                   float64
gameId                       int64
gameDateTimeEst                str
playerteamCity                 str
playerteamName                 str
opponentteamCity               str
opponentteamName               str
gameType                       str
gameLabel                      str
gameSubLabel                   str
seriesGameNumber            object
win                        float64
home                       float64
numMinutes                  object
points                     float64
assists                    float64
blocks                     float64
steals                     float64
fieldGoalsAttempted        float64
fieldGoalsMade             float64
fieldGoalsPercentage       float64
threePointersAttempted     float64
threePointersMade          float64
threePointersPercentage    float64
freeThrowsAttempted        float64
freeThrowsMade             float64
freeThr

In [18]:
print('Первые 5 строк:')
df.head()

Первые 5 строк:


,firstName,lastName,personId,gameId,gameDateTimeEst,playerteamCity,playerteamName,opponentteamCity,opponentteamName,gameType,...,reboundsOffensive,reboundsTotal,foulsPersonal,turnovers,plusMinusPoints,playerteamId,opponentteamId,comment,startingPosition,gameDate
0,Josh,Hart,1628404.0,42500405,2026-06-13 20:30:00,New York,Knicks,San Antonio,Spurs,Playoffs,...,3.0,11.0,3.0,1.0,15.0,1.610613e+09,1.610613e+09,NaN,F,2026-06-13 20:30:00
1,OG,Anunoby,1628384.0,42500405,2026-06-13 20:30:00,New York,Knicks,San Antonio,Spurs,Playoffs,...,0.0,8.0,5.0,1.0,2.0,1.610613e+09,1.610613e+09,NaN,F,2026-06-13 20:30:00
2,Jeremy,Sochan,1631110.0,42500405,2026-06-13 20:30:00,New York,Knicks,San Antonio,Spurs,Playoffs,...,0.0,0.0,0.0,0.0,1.0,1.610613e+09,1.610613e+09,NaN,NaN,2026-06-13 20:30:00
3,Jordan,McLaughlin,1629162.0,42500405,2026-06-13 20:30:00,San Antonio,Spurs,New York,Knicks,Playoffs,...,0.0,0.0,0.0,0.0,0.0,1.610613e+09,1.610613e+09,DNP - Coach's Decision,NaN,2026-06-13 20:30:00
4,Jordan,Clarkson,203903.0,42500405,2026-06-13 20:30:00,New York,Knicks,San Antonio,Spurs,Playoffs,...,1.0,1.0,0.0,0.0,3.0,1.610613e+09,1.610613e+09,NaN,NaN,2026-06-13 20:30:00


In [19]:
print('Пропущенные значения по колонкам:')
print(df.isnull().sum())

Пропущенные значения по колонкам:
firstName                      234
lastName                       238
personId                        45
gameId                           0
gameDateTimeEst               3513
playerteamCity                 168
playerteamName                3513
opponentteamCity              3681
opponentteamName              3513
gameType                        50
gameLabel                  1552474
gameSubLabel               1564831
seriesGameNumber           1532465
win                           3523
home                          3513
numMinutes                  168636
points                        1687
assists                       1687
blocks                        1687
steals                        1687
fieldGoalsAttempted           1687
fieldGoalsMade                1687
fieldGoalsPercentage          1687
threePointersAttempted        1687
threePointersMade             1687
threePointersPercentage       1687
freeThrowsAttempted           1687
freeThrowsMade       

In [20]:

rows_before = len(df)

df_clean = df.copy()

# убираем строки где points пустые
df_clean = df_clean.dropna(subset=['points'])

# убираем строки где numMinutes = 0 или пустые
df_clean = df_clean.dropna(subset=['numMinutes'])
df_clean = df_clean[df_clean['numMinutes'] != 0]

# преобразуем дату и извлекаем год
df_clean['gameDateTimeEst'] = pd.to_datetime(df_clean['gameDateTimeEst'])
df_clean['Season_Year'] = df_clean['gameDateTimeEst'].dt.year

rows_after = len(df_clean)
lost = rows_before - rows_after
pct_lost = lost / rows_before * 100

print(f'Строк до очистки:  {rows_before}')
print(f'Строк после:       {rows_after}')
print(f'Потеряно:          {lost} ({pct_lost:.1f}%)')


Строк до очистки:  1669922
Строк после:       1409415
Потеряно:          260507 (15.6%)


In [23]:
df_clean['Decade'] = (df_clean['Season_Year'] // 10 * 10).astype(str) + 'е'

decade_stats = (
    df_clean
    .groupby('Decade', sort=True)
    .agg(
        Avg_Points   =('points',        'mean'),
        Avg_Assists  =('assists',        'mean'),
        Avg_Rebounds =('reboundsTotal',  'mean'),
        Count        =('points',         'count'),
    )
    .round(2)
    .reset_index()
)

decade_stats
#Больше всего очков игроки набирали в 1960-е — в среднем 12.4 очка за игру, это пик за всю историю. После этого результативность постепенно снижалась вплоть до 2010-х годов, где  
  #опустилась до минимума — около 9.6 очка. Самый резкий обрыв произошёл на переходе от 1960-х к 1970-м: тогда НБА резко расширилась — появилось много новых команд, и места в       
  #составах заняли менее результативные игроки, что «разбавило» общую статистику. В 2020-х виден небольшой рост обратно до 10.5 — это эпоха «трёхочкового бума», когда вся лига      
  #начала активнее атаковать из-за дуги.

,Decade,Avg_Points,Avg_Assists,Avg_Rebounds,Count
0,1940.0е,7.62,0.04,0.00,3511
1,1950.0е,9.84,1.35,3.01,17722
2,1960.0е,12.44,2.34,6.33,44596
3,1970.0е,11.03,2.50,4.94,148567
4,1980.0е,10.72,2.53,4.29,203517
5,1990.0е,10.00,2.31,4.19,235822
6,2000.0е,9.61,2.11,4.15,273255
7,2010.0е,9.72,2.11,4.09,289751
8,2020.0е,10.52,2.38,4.09,189629


In [24]:
df_clean['numMinutes'] = pd.to_numeric(df_clean['numMinutes'], errors='coerce')
df_clean['points']    = pd.to_numeric(df_clean['points'],    errors='coerce')

player_stats = (
    df_clean
    .groupby(['personId', 'firstName', 'lastName'], as_index=False)
    .agg(
        Total_Points  = ('points',     'sum'),
        Total_Minutes = ('numMinutes', 'sum'),
        Games         = ('points',     'count'),
    )
)
player_stats['Player'] = player_stats['firstName'] + ' ' + player_stats['lastName']

# Топ-20 по общим очкам
top20_total = (
    player_stats
    .nlargest(20, 'Total_Points')
    [['Player', 'Total_Points', 'Games']]
    .reset_index(drop=True)
)
top20_total.index += 1

# Топ-20 по очкам за 36 минут (минимум 100 игр)
player_stats['Pts_per_36'] = (player_stats['Total_Points'] / player_stats['Total_Minutes'] * 36).round(2)
top20_per36 = (
    player_stats[player_stats['Games'] >= 100]
    .nlargest(20, 'Pts_per_36')
    [['Player', 'Pts_per_36', 'Games']]
    .reset_index(drop=True)
)
top20_per36.index += 1

# Таблицы рядом
side_by_side = pd.concat(
    [top20_total.add_prefix('Total__'), top20_per36.add_prefix('Per36__')],
    axis=1
)
print('=== Топ-20: общие очки vs очки за 36 минут ===\n')
print(side_by_side.to_string())

# Сколько игроков совпадает
overlap = set(top20_total['Player']) & set(top20_per36['Player'])
print(f'\nСовпадений в обоих списках: {len(overlap)}')
if overlap:
    print('Игроки:', ', '.join(sorted(overlap)))


=== Топ-20: общие очки vs очки за 36 минут ===

          Total__Player  Total__Total_Points  Total__Games     Per36__Player  Per36__Pts_per_36  Per36__Games
1          LeBron James              53845.0          2037      Arnie Ferrin                inf           119
2   Kareem Abdul-Jabbar              43286.0          1775        Tony Jaros                inf           107
3           Karl Malone              41745.0          1676         Paul Noel                inf           108
4           Kobe Bryant              40452.0          1633       Jack Kerris             774.00           124
5          Kevin Durant              39324.0          1466     Arnie Johnson             454.74           161
6        Michael Jordan              38415.0          1258   Thomas Eddleman             426.25           105
7         Dirk Nowitzki              36342.0          1747          Al Cervi             370.92           104
8          James Harden              34988.0          1495     David Joh

## Выводы

**Совпадений: 0 — почему так мало**

Два списка отбирают игроков из разных эпох. Топ по общим очкам — это звёзды с длинными карьерами (LeBron, Kareem, Kobe): они набрали много просто потому что играли 15–20 лет. Топ по Pts/36 оказался полностью забит игроками **1940-х годов** — и это сигнал о проблеме с данными: у многих из них минуты либо не записывались вообще (получается деление почти на ноль → `inf`), либо записаны некорректно. Поэтому пересечение нулевое — списки смотрят в разные стороны.

---

**Что такое Pts_per_36**

Это очки, которые игрок набрал бы за стандартные 36 минут игры — если бы играл с той же скоростью что и обычно. Метрика убирает несправедливость: резервный игрок выходит на 12 минут и набирает 8 очков — это лучше, чем кажется. `Pts_per_36 = Total_Points / Total_Minutes × 36`. В современной НБА это стандартный способ сравнивать игроков с разным игровым временем.

---

**Кто лучший игрок**

По общим очкам — бесспорно **LeBron James** (53 845 очков, 2037 игр). Но если смотреть на эффективность — **Michael Jordan**: 38 415 очков всего за 1258 игр, то есть ~30.5 очка за игру. Это лучшее соотношение среди игроков топ-20 по тотальным очкам в современной эре. Топ по Pts/36 для 1940-х доверять не стоит — там проблема с качеством данных о минутах.


In [25]:
season_stats = (
    df_clean
    .groupby('Season_Year', as_index=False)
    .agg(
        Avg_Points   = ('points',    'mean'),
        Num_Players  = ('personId',  'nunique'),
    )
    .round({'Avg_Points': 2})
)

top10_seasons = (
    season_stats
    .nlargest(10, 'Avg_Points')
    .reset_index(drop=True)
)
top10_seasons.index += 1

print('=== Топ-10 самых результативных сезонов ===\n')
print(top10_seasons.to_string())

print('\n--- Все сезоны (по убыванию среднего) ---\n')
print(season_stats.sort_values('Avg_Points', ascending=False).to_string(index=False))


=== Топ-10 самых результативных сезонов ===

    Season_Year  Avg_Points  Num_Players
1        1960.0       14.19          104
2        1961.0       13.95          120
3        1962.0       13.31          140
4        1959.0       13.18          100
5        1967.0       12.74          164
6        1966.0       12.54          133
7        1968.0       12.28          188
8        1969.0       12.23          197
9        1965.0       12.10          132
10       1963.0       11.91          127

--- Все сезоны (по убыванию среднего) ---

 Season_Year  Avg_Points  Num_Players
      1960.0       14.19          104
      1961.0       13.95          120
      1962.0       13.31          140
      1959.0       13.18          100
      1967.0       12.74          164
      1966.0       12.54          133
      1968.0       12.28          188
      1969.0       12.23          197
      1965.0       12.10          132
      1963.0       11.91          127
      1970.0       11.88          229
    

## Вывод

Все десять самых результативных сезонов в истории НБА приходятся на **1960-е годы** — лидер 1961 год с 12.49 очка в среднем за игру. Это эпоха до введения строгих правил защиты и активного расширения лиги: команд было мало, игроки звёздного уровня выходили на площадку почти без замен и набирали огромное количество очков. После 1970 года результативность начала падать по мере роста числа игроков и усиления защитной тактики — и больше к уровню 1960-х так и не вернулась.


# 🏀 NBA Statistics — Итоги расследования

## Данные
- Исходный датасет: **1 669 922 строки** — после удаления пустых очков и нулевых минут осталось **1 409 415 строк** (потеряли 15.6%)

## Открытие 1 — Результативность по эпохам
- Лучшее десятилетие — **1960-е: 12.44 очка за игру**, хотя интуитивно кажется что современные игроки результативнее — на деле сегодня в среднем лишь 10.5 очка

## Открытие 2 — Лучшие сезоны
- Все топ-10 самых результативных сезонов — это **1960–1970 годы**, когда в лиге играло всего 110–200 игроков: маленький пул звёзд без ротации гнал статистику вверх

## Открытие 3 — Звёзды или система?
- В топ-20 по общим очкам и топ-20 по очкам за 36 минут **не совпал ни один игрок из 40** — это два принципиально разных рейтинга, измеряющих разные вещи

## ⚠️ Пойманные ловушки
- **Ловушка 1:** Общие очки за карьеру — это рейтинг долголетия, а не таланта. LeBron James на первом месте (53 845 очков) просто потому что провёл в НБА 21 сезон — Michael Jordan набрал почти столько же за 15 лет при среднем 30.5 очка за игру
- **Ловушка 2:** 1960-е выглядят эпохой супергероев, но тогда в лиге было в 5–6 раз меньше игроков чем сейчас — среднее завышено потому что в статистику попадали только настоящие звёзды, а не весь состав из 15 человек

## 🏆 Главное открытие для презентации
- **0 совпадений из 40** — ни один игрок не попал одновременно в топ по общим очкам и топ по эффективности за 36 минут: это доказывает что стандартные рейтинги НБА измеряют выносливость карьеры, а не качество игры
